test/types.test.ts

```diff
@@ -31,7 +36,20 @@ export const testPluginWithAsync = fp(
   }
 );
 
+// Register with HTTP
 const server = fastify()
 
-server.register(testPlugin) // register expects a fastify.Plugin
-server.register(testPluginWithCallback) // test fastify.Plugin's gerneric typings
+server.register(testPluginWithOptions) // register expects a FastifyPlugin
+server.register(testPluginWithCallback)
+server.register(testPluginWithAsync)
+
+// Register with HTTP2
+const serverWithHttp2: FastifyInstance<
```

Comment:
You shouldn't need to specify these generics any longer. Let me know if removing them causes issues and I'll jump in and try and figure out whats up.

Reply:
I don't need to specify the generics but if I remove them I get HTTP types, not HTTP2 types.

I think we might need to look at the types themselves. 

When authoring a plugin, we will not know what generics will be passed into `FastifyInstance` so maybe the types need to cater for that? 
e.g. `FastifyPlugin<Server = http.Server & https.Server & http2.Http2Server>`

The other issue I find strange is that server type is determined by fastify options. If https options are passed in, the server is of type https. If http2 boolean is passed in, the server is of that type. Shouldn't the types reflect that mechanism?

Reply:
Not exactly, the server types are *supposed* to be inferred based on the options object. The generics only exist if you want to extend the types of those servers, request, or responses. See here:

https://github.com/fastify/fastify/blob/f1cfcadbe201f98137a5b2a9ee1159215b5d81e5/fastify.d.ts#L20-L44

This is how the `fastify()` method is defined in the `next` types (if you're curious its using discriminant unions and function overloads to achieve it).

Fastify-Plugin should forward the generic properties through to the underlying instance, and it should also enable the user to omit them if they are not decorating those generic parameters

Reply:
You're right, the server type is correctly inferred, that's great!

I'm still not clear how the `FastifyPlugin` type is supposed to be used though.

Plugins are defined like so:

```ts
export const testPlugin: FastifyPlugin = fp(
  function (fastify, _options, _next) {
    console.log(fastify) // Is now of type `FastifyInstance<Server, IncomingMessage, ServerResponse, FastifyLoggerOptions<Server, IncomingMessage, ServerResponse>>`
  },
  {},
)
```

How will TypeScript know what options have been passed into `fastify`? The fastify instance hasn't even been created at that point.

Consequently in user land, the following line is failing because of the defaults used by `FastifyPlugin` type:

```ts
const serverWithHttp2 = fastify({ http2: true });
serverWithHttp2.register(testPlugin) // Argument of type 'FastifyPlugin<{}, Server, IncomingMessage, ServerResponse>' is not assignable to parameter of type 'FastifyPlugin<{}, Http2Server, Http2ServerRequest, Http2ServerResponse>'.ts(2345)
```

Can you explain how to make this work correctly?

Reply:
Hmm this is a touch challenge. Let me check this out locally and start investigating.

Reply:
@mcollina can you foresee any circumstance where a plugin should only work for a specific type (http vs https vs http2) of fastify server? Disregarding TypeScript, is there anything currently in place that validates that a (for example) http2-only plugin is being registered in a http2 fastify instance?

Reply:
@thomheymann after working with this for a little bit I believe we are trying to solve a problem that doesn't exist (_yet_). As far as I'm aware plugins aren't scoped to a specific type of server instance and as such shouldn't have to pass some type check for that either. I was able to resolve the issue you're facing here by removing the defaults in the FastifyPlugin definition in the fastify type defs. This still enforces that the server, request, and response all match, but then allows for any plugin to be registered to any server instance. I'm going to wait Matteo's reply and then will proceed as I have other ideas for solutions depending on what he says.

Reply:
@mcollina can you foresee any circumstance where a plugin should only work for a specific type (http vs https vs http2) of fastify server? 

No, the vast majority of plugins should infer the http vs https vs http2 thing automatically. I _thought_ this was what v3 types were about :/.

Reply:
@Ethan-Arrowood  Can you share your solution? When I remove the defaults I get typing errors asking me to provide types for the generics

Reply:
Okay I'll work on a response soon. Its sorta complicated. 

Short answer is this:
>  I was able to resolve the issue you're facing here by removing the defaults in the FastifyPlugin definition in the fastify type defs

Without modifying Fastify types you can cast a plugin to `unknown` then cast it to your intended FP instance inline which is a bit better that specifying with generic parameters. Also, the answer might be to just use generic parameter inference. Its not a great solution but it definitely is one. 

So like I said let me spend some more time with this and ill write back as soon as I can!

Reply:
I'm not finished yet, but I'm starting to simplify the problem in a [TypeScript playground](https://www.typescriptlang.org/play/index.html?ssl=2&ssc=22&pln=1&pc=1#code/JYWwDg9gTgLgBACxjMBnOAzKERwORIqp4BQoksiyYm2uB1p508hYATLTvm+6STACeYAKZwA8mBjAIAO3QBeOAG8ScdXAA2EAOY6RUAPwAuOACMIETSICGskgF8SA4WIAS1VJOlzFEqTLycABkKmoabKimkQB0AMoGAG4G3oGojs5ConAeKOypvnBKBUGhqhpUeaYwUACuIhku2bkcCQDGtVAiJX49IWEVvNV1IgA04eqR0dTs8SIdXQlQyVA9jVliSysAkrIwBqjzPrJFlWBzywZwAD5nqBcrN2ezLexbV7e8Ma-tnSLvUGcwD2BgwNjaYgAYjZUNIMIIAAqaWo6YEAHnCPXG6gBcBEAA99rIACboAG7fZQQ5tY4kAB8Aw0jWBlLBELg0NhwHhACURKjYQYMTiklcCUTSXBySCqUdAvTGeo0T06QAKCYVOBgZGo2SmTlwxE69FYqWiqB07Ga9QQKSoEz+Y7oMparoYYD4h2wqDAnQAbi0ugAMiJkpovTVfXAnBUAJSmRIQYDEv2NMgytlQmGG3awuBo3HikQksnmikHOVyBnlDSHS5QHk2ADupgBqYqXQFlNMjQwtVkNMCmGz3MEwo0hcJxclXx+8z+APpqttMCiORmvy6PXjHJH8NzMAL5rpJD7A+Ow658PHIvreKnJee3w35qXK7Xr23+r3ggPR-rJ5noOciXoaN5mneRaPrEi5qu+pgtF4AS+DuBqjn+AKAf2wEnGCV5jhqk4SugbAPAYb52g6X67vhGHHqe2EXnhYFEdOpb1uWso4WqGrvlRyFBLciF9EJMwieueSbt0AnpKhP50RxMrUsc1YatY8B1is7YaNycCqgAhO+saKpqmlXEopFtF0Nj7ACqqxhqDh4pohxwLpqoAESRB5bknO+IShPpnm8D5wJwEZxk1taZlQKcsRWbYtnmsudoxJEDkVE5IguWIbkYHpXkzKFfl2gFcBBYVRDFeFdqxpFGoVDFcUzDECU2f8yUZRoWU5eo7mVRw1X+cEoQDagQ21SZjXms1eStdZSULB19YpauaWeF16g9a5UWmTNFnUPNiXLSs9mOc4HYiDAnQnLt031o2LZwDFVrWp2wCClAPaZRqjhAA)

I'm super busy with work right now and have to switch back to that for the time being, but I'll keep working on this tonight if I can 😄

Reply:
As I continue working on my playground example I wanted to share something I'm pretty certain about.

One of the challenges here is that when you create a fastify plugin there is no way to declare that it is for any specific server type or set of server types. This is even a limitation in JavaScript; the logic for if a plugin is locked into http, https, or http2 would have to exist within the plugin code itself (i.e. some top level check for the right server type). This mean in TypeScript land we can only safely type a fastify plugin as "it can be used for any server type" too. A TS dev can utilize type guarding and casting within their function if they need to utilize something specific about the fastify instance, just like a JS dev would need to check for the existence of such properties or methods. 

What I can derive from this is maybe we consider adding some new options to the fastify-plugin thing that allows us to specify what server type a plugin should be for; this would then enable the TS side to use discriminant unions and overloads to make plugins type safe based on their options object.  

Your idea @thomheymann might be a great option. I'm going to try this out and see what the dev experience is like.
> When authoring a plugin, we will not know what generics will be passed into FastifyInstance so maybe the types need to cater for that?
e.g. FastifyPlugin<Server = http.Server & https.Server & http2.Http2Server>

Reply:
My point of view is that developers should not have to specify any generics for plugins, and cast things if they need to. Essentially, fastify-plugin should be pass-through.

Reply:
[Here](https://www.typescriptlang.org/play/index.html?ssl=126&ssc=1&pln=134&pc=1#code/JYWwDg9gTgLgBACxjMBnOAzKERwORIqp4BQoksiyYm2uB1p508hYATLTvm+6STACeYAKZwA8mBjAIAO3QBeOAG8ScdXAA2EAOY6RUAPwAuOACMIETSICGskgF8SA4WIAS1VJOlzFEqTLycABkKmoabKimkQB0AMoGAG4G3oGojs5ConAeKOypvnBKBUGhqhpUeaYwUACuIhku2bkcCQDGtVAiJX49IWEVvNV1IgA04eqR0dTs8SIdXQlQyVA9jVliSysAkrIwBqjzPrJFlWBzywZwAD5nqBcrN2ezLexbV7e8Ma-tnSLvUGcwD2BgwNjaYgAYjZUNIMIIAAqaWo6YEAHnCPXG6gBcBEAA99rIACboAG7fZQQ5tY6nNgPAwkAB8Aw0AAoJhU4MDYXYIaZobDgPDdrC0QCmdiueoIFIov5jqgpdLZASYKY2QBKIosxIQYDEzma0x6g0AbkaGFqshpgTgkIRGIqWM5uLVIhJZKSBgpByOgWZHIqGFkAphcMRyNRsjRWLgEuVcFlMFQJgVaUTqsJaa1OrgpsNxvt4eFkZR6LjEtZ6i6ME6JxDjWBlLBEOLQvhACURKjYQYnfHvVA8YSPaTB5coL6qf65HTqAzASzyhpYwFfEyg9K4GAo8Cwx2y9G14rRhOVpLORVk6nTH0yjuuhhgPi07CoMCdGatLoADIiZJNDfGpPzgJwKiLAsLScMgQSgVsoRLEV5BgcUhxHIlx3JODqWOZdOUOSdOxsAB3UwAQtCoul7SkDwjbsaP7CVKI0Xdy3kOjSyRdi0TsQQzwlABtABdaDnCtG1aTBQ8BxxdD3U9Z5vhmX5FiHQMb1MH55j+Hoi0FCNRVQiUSAk2052kiNZPPK4FPHL5tMnDS5S0mY9M45CxRMsypKQwRrLdUdFNiAFnJTVyiHc9tDJQtDJyZUzrXMhs-IC+SgvshdQs3G80yigzSyMuKL0SyS7Us0titsjKvUnadcMCTdOVyu91yCW4Wi8Nr0A6tzuqebSFm6bqSH0vyiuwykGrkfCKmseBCJWFj1GFOA2QAQhvbUV2lRariUek2i6Gx9gBLVOQcPFNEObkMDWgAiSJ7u5E4bxCUJ1rZR6Zme4EkzlTVtqvDQ9uHA7PBiI7bFOoc2RvGJIk1C6rpu26Ht4X7Xrld64E+76iEx-6U0B6suVB+c8kh46Ycnc6KkukRrrENGvoxl6ifQYJQnxtBCa20mKnJ8HKahk7-h0tTafhxHkcZm6drJ9DhfOUWaZWOmNBgio2l8eA2OjeUCvhbjjz4gShyZETThE5xtd1uBqOAPsoA8wQGKdykqqgFklDZfXgW1BRZu3f35BiMBalQBA-b3WQkfp22NFresBZBodiLIuBQcTKiew9gwc9Y2P0gTmCdZQncAEZXZN9FlAcM96SrX2eRgM8bzPLMYED5cnHL2Ed3YGvY7RevG4h5u1tb9u5U7tUe5UPv7bAABmYeeLHpTHIvU42Wnjm58JBf65Ifu9YAFnX49N4clSJfF+Ld-3ju4C74+y-thFK9OB0-crpGz5wARJwJQv8OAAM-ivH+CI-YrwgRXBE59oF+3PkjU+9s2C4iUBVeE51MFDhiI7Z2bIv5I3wZOQhediHALIdQAElDGJQBIXAkg5CVgMPzkwxBaDAGRCwZgPybJlB3FMPXMCtCiD0KIZSEh-9WGeCkVQmRND5GSIIdIgwzCJFoEUYwkhqDnC8LvpOU4ODBBCOeMMeo4jVGtHUUozRpDbFvHsXolRvBdGcK0c4zx1CDHoIrh4++-CzEWKGHAGo9Rx5EFEZdBw2iXFDV8couRQSkmuK8e4u+6SKEaK4SwtJfxkmOIMUAA) is my latest playground link. 

This is a tough issue; I'm having a hard time landing on some resolution. **What do you feel like is the best developer experience for TypeScript developers here?** I need more opinions so we can zero-in on a solution

Reply:
Based on @mcollina answer I believe the `FastifyPlugin` interface in the `fastify` repo should remove the generics `RawServer`, `RawRequest`, and `RawResponse`. Also removing them from this repo.

If the user needs a specific server instance in their plugin they need to use type casting or type guarding

Reply:
Can we try that?

Reply:
@thomheymann do you want to try making the changes I recommended?

Reply:
Would it be possible to just leave the `FastifyPlugin` instance type as the `ServerIntersection` by default? It reflects the reality of plugin authors not knowing what exact server type will be used. For those plugins that depend on a specific server or modified request/response objects those should be able to be passed in but most plugins will be generic. 

I've had a stab here at this here:
https://www.typescriptlang.org/play/index.html?ssl=113&ssc=2&pln=108&pc=1#

* I mocked http/https/http2 namespaces as the playground couldn't find dependencies
* `FastifyPlugin` instance type is now union of all server types (its called intersection but it's a union, worst naming in typescript ever):

    ```ts
    interface FastifyPlugin<Options, Server extends ServerIntersection = ServerIntersection> {
        (
            instance: FastifyInst<ServerIntersection>,
            opts: Options,
            next: () => void
        ): void;
    }
    ```

* Plugin authors get correct instance type:

    ```ts
    // Generic plugin
    const p4: FastifyPlugin<{}> = (inst, opts, next) => {
        console.log(inst.serverRaw.type) // "http" | "https" | "http2" | "https2"
    }

    // Specific plugin
    const p3: FastifyPlugin<{}, https.Server> = (inst, opts, next) => {
        console.log(inst.serverRaw.type) // "https"
    }
    ```

* Userland Fastify instance type is correctly inferred based on options
* Userland plugin registration works without typing errors

Reply:
Seems a good idea to me but my TS is far from good.

Reply:
Yeah this looks good. I'm happy shipping this now and if things change in the future we can of course change it ourselves.

Yeah this is one of those edge cases that is really hard to solve correctly because it wouldn't happen when using TS from the ground up. If I have the opportunity I'll pass this along to the TS team for consideration.

Reply:
This PR has been updated but will fail until corresponding Fastify PR has been merged in:
https://github.com/fastify/fastify/pull/2183